File: packages/identity-engine/src/calibration.ts
import type {
  CalibrationMetrics,
  CalibrationObjective,
  CalibrationSample,
  IdentityScenario,
  IdentityType,
  ThresholdCalibrationProfile,
} from "./types";

function assertProbability(value: number, name: string): void {
  if (!Number.isFinite(value) || value < 0 || value > 1) throw new Error(`${name} must be between 0 and 1`);
}

function assertScore(value: number): void {
  if (!Number.isFinite(value) || value < 0 || value > 100) throw new Error("calibration score must be between 0 and 100");
}

function countsAt(samples: readonly CalibrationSample[], threshold: number) {
  let tp = 0;
  let fp = 0;
  let tn = 0;
  let fn = 0;
  for (const sample of samples) {
    const positive = sample.label === "POSITIVE";
    const predicted = sample.score >= threshold;
    if (positive && predicted) tp += 1;
    else if (positive) fn += 1;
    else if (predicted) fp += 1;
    else tn += 1;
  }
  return { tp, fp, tn, fn };
}

function safeDivide(numerator: number, denominator: number): number {
  return denominator === 0 ? 0 : numerator / denominator;
}

function rocAuc(samples: readonly CalibrationSample[]): number {
  const positives = samples.filter((sample) => sample.label === "POSITIVE");
  const negatives = samples.filter((sample) => sample.label !== "POSITIVE");
  if (!positives.length || !negatives.length) return 0;
  let wins = 0;
  for (const positive of positives) {
    for (const negative of negatives) {
      if (positive.score > negative.score) wins += 1;
      else if (positive.score === negative.score) wins += 0.5;
    }
  }
  return wins / (positives.length * negatives.length);
}

function averagePrecision(samples: readonly CalibrationSample[]): number {
  const sorted = [...samples].sort((a, b) => b.score - a.score || a.sample_id.localeCompare(b.sample_id));
  const positives = sorted.filter((sample) => sample.label === "POSITIVE").length;
  if (!positives) return 0;
  let seenPositive = 0;
  let precisionSum = 0;
  sorted.forEach((sample, index) => {
    if (sample.label !== "POSITIVE") return;
    seenPositive += 1;
    precisionSum += seenPositive / (index + 1);
  });
  return precisionSum / positives;
}

export function selectCalibratedThreshold(
  samples: readonly CalibrationSample[],
  identityType: IdentityType,
  scenario: IdentityScenario,
  objective: CalibrationObjective = {},
): CalibrationMetrics {
  const filtered = samples.filter((sample) => sample.identity_type === identityType && sample.scenario === scenario);
  if (!filtered.length) throw new Error("calibration dataset is empty for identity type/scenario");
  filtered.forEach((sample) => assertScore(sample.score));
  const positiveCount = filtered.filter((sample) => sample.label === "POSITIVE").length;
  const negativeCount = filtered.filter((sample) => sample.label === "NEGATIVE").length;
  const nearMissCount = filtered.filter((sample) => sample.label === "NEAR_MISS").length;
  if (!positiveCount || !(negativeCount + nearMissCount)) {
    throw new Error("calibration requires positive and negative/near-miss samples");
  }
  const minimumPrecision = objective.minimum_precision ?? 0;
  const minimumRecall = objective.minimum_recall ?? 0;
  assertProbability(minimumPrecision, "minimum_precision");
  assertProbability(minimumRecall, "minimum_recall");

  const thresholds = [...new Set(filtered.map((sample) => sample.score))].sort((a, b) => a - b);
  let best: { threshold: number; precision: number; recall: number; f1: number; fpr: number; fnr: number } | null = null;
  for (const threshold of thresholds) {
    const { tp, fp, tn, fn } = countsAt(filtered, threshold);
    const precision = safeDivide(tp, tp + fp);
    const recall = safeDivide(tp, tp + fn);
    if (precision < minimumPrecision || recall < minimumRecall) continue;
    const f1 = safeDivide(2 * precision * recall, precision + recall);
    const fpr = safeDivide(fp, fp + tn);
    const fnr = safeDivide(fn, fn + tp);
    const candidate = { threshold, precision, recall, f1, fpr, fnr };
    if (
      !best ||
      candidate.f1 > best.f1 ||
      (candidate.f1 === best.f1 && candidate.precision > best.precision) ||
      (candidate.f1 === best.f1 && candidate.precision === best.precision && candidate.recall > best.recall) ||
      (candidate.f1 === best.f1 && candidate.precision === best.precision && candidate.recall === best.recall &&
        ((objective.prefer_higher_threshold_on_tie ?? true) ? candidate.threshold > best.threshold : candidate.threshold < best.threshold))
    ) {
      best = candidate;
    }
  }
  if (!best) throw new Error("no calibrated threshold satisfies objective");
  return {
    threshold: best.threshold,
    precision: best.precision,
    recall: best.recall,
    f1: best.f1,
    false_positive_rate: best.fpr,
    false_negative_rate: best.fnr,
    roc_auc: rocAuc(filtered),
    average_precision: averagePrecision(filtered),
    positive_count: positiveCount,
    negative_count: negativeCount,
    near_miss_count: nearMissCount,
  };
}

export interface BuildCalibrationProfileInput {
  readonly profile_id: string;
  readonly organization_id: string;
  readonly identity_type: IdentityType;
  readonly scenario: IdentityScenario;
  readonly version: string;
  readonly model_bundle_version: string;
  readonly preprocessor_version: string;
  readonly calibration_dataset_version: string;
  readonly signal_weights: Readonly<Record<string, number>>;
  readonly required_signals: readonly string[];
  readonly review_margin: number;
  readonly minimum_confidence: number;
  readonly samples: readonly CalibrationSample[];
  readonly objective?: CalibrationObjective;
}

export function buildCalibrationProfile(input: BuildCalibrationProfileInput): ThresholdCalibrationProfile {
  if (!input.required_signals.length) throw new Error("required_signals must not be empty");
  const distinctSignals = new Set(input.required_signals);
  if ((input.identity_type === "PRODUCT" || input.identity_type === "LOGO") && distinctSignals.size < 2) {
    throw new Error("PRODUCT/LOGO calibration must require multiple independent signals");
  }
  let totalWeight = 0;
  for (const [signal, weight] of Object.entries(input.signal_weights)) {
    if (!Number.isFinite(weight) || weight <= 0) throw new Error(`signal weight must be positive: ${signal}`);
    totalWeight += weight;
  }
  if (totalWeight <= 0) throw new Error("signal_weights must not be empty");
  for (const signal of input.required_signals) {
    if (!(signal in input.signal_weights)) throw new Error(`required signal has no weight: ${signal}`);
  }
  assertProbability(input.minimum_confidence, "minimum_confidence");
  if (!Number.isFinite(input.review_margin) || input.review_margin < 0 || input.review_margin > 100) {
    throw new Error("review_margin must be between 0 and 100");
  }
  const metrics = selectCalibratedThreshold(input.samples, input.identity_type, input.scenario, input.objective);
  return {
    profile_id: input.profile_id,
    organization_id: input.organization_id,
    identity_type: input.identity_type,
    scenario: input.scenario,
    version: input.version,
    status: "PUBLISHED",
    threshold: metrics.threshold,
    review_floor: Math.max(0, metrics.threshold - input.review_margin),
    minimum_confidence: input.minimum_confidence,
    signal_weights: { ...input.signal_weights },
    required_signals: [...distinctSignals].sort(),
    model_bundle_version: input.model_bundle_version,
    preprocessor_version: input.preprocessor_version,
    calibration_dataset_version: input.calibration_dataset_version,
    metrics,
  };
}

File: services/identity-engine/src/lumi_identity_engine/calibration.py
from __future__ import annotations

from collections.abc import Iterable

from .model import (
    CalibrationMetrics,
    CalibrationSample,
    IdentityScenario,
    IdentityType,
    ThresholdCalibrationProfile,
)


def _divide(numerator: float, denominator: float) -> float:
    return 0.0 if denominator == 0 else numerator / denominator


def _roc_auc(samples: tuple[CalibrationSample, ...]) -> float:
    positives = tuple(row for row in samples if row.label == "POSITIVE")
    negatives = tuple(row for row in samples if row.label != "POSITIVE")
    if not positives or not negatives:
        return 0.0
    wins = 0.0
    for positive in positives:
        for negative in negatives:
            if positive.score > negative.score:
                wins += 1.0
            elif positive.score == negative.score:
                wins += 0.5
    return wins / (len(positives) * len(negatives))


def _average_precision(samples: tuple[CalibrationSample, ...]) -> float:
    ordered = sorted(samples, key=lambda row: (-row.score, row.sample_id))
    positive_count = sum(row.label == "POSITIVE" for row in ordered)
    if positive_count == 0:
        return 0.0
    seen_positive = 0
    precision_sum = 0.0
    for index, row in enumerate(ordered, start=1):
        if row.label == "POSITIVE":
            seen_positive += 1
            precision_sum += seen_positive / index
    return precision_sum / positive_count


def select_calibrated_threshold(
    samples: Iterable[CalibrationSample],
    identity_type: IdentityType,
    scenario: IdentityScenario,
    *,
    minimum_precision: float = 0.0,
    minimum_recall: float = 0.0,
) -> CalibrationMetrics:
    rows = tuple(
        row for row in samples if row.identity_type == identity_type and row.scenario == scenario
    )
    if not rows:
        raise ValueError("calibration dataset is empty for identity type/scenario")
    if not 0 <= minimum_precision <= 1 or not 0 <= minimum_recall <= 1:
        raise ValueError("calibration objective must be between 0 and 1")
    for row in rows:
        if not 0 <= row.score <= 100:
            raise ValueError("calibration score must be between 0 and 100")
    positive_count = sum(row.label == "POSITIVE" for row in rows)
    negative_count = sum(row.label == "NEGATIVE" for row in rows)
    near_miss_count = sum(row.label == "NEAR_MISS" for row in rows)
    if positive_count == 0 or negative_count + near_miss_count == 0:
        raise ValueError("calibration requires positive and negative/near-miss samples")

    best: tuple[float, float, float, float, float, float] | None = None
    for threshold in sorted({row.score for row in rows}):
        tp = fp = tn = fn = 0
        for row in rows:
            expected = row.label == "POSITIVE"
            predicted = row.score >= threshold
            if expected and predicted:
                tp += 1
            elif expected:
                fn += 1
            elif predicted:
                fp += 1
            else:
                tn += 1
        precision = _divide(tp, tp + fp)
        recall = _divide(tp, tp + fn)
        if precision < minimum_precision or recall < minimum_recall:
            continue
        f1 = _divide(2 * precision * recall, precision + recall)
        fpr = _divide(fp, fp + tn)
        fnr = _divide(fn, fn + tp)
        candidate = (f1, precision, recall, threshold, fpr, fnr)
        if best is None or candidate[:4] > best[:4]:
            best = candidate
    if best is None:
        raise ValueError("no calibrated threshold satisfies objective")
    f1, precision, recall, threshold, fpr, fnr = best
    return CalibrationMetrics(
        threshold=threshold,
        precision=precision,
        recall=recall,
        f1=f1,
        false_positive_rate=fpr,
        false_negative_rate=fnr,
        roc_auc=_roc_auc(rows),
        average_precision=_average_precision(rows),
        positive_count=positive_count,
        negative_count=negative_count,
        near_miss_count=near_miss_count,
    )


def build_calibration_profile(
    *,
    profile_id: str,
    organization_id: str,
    identity_type: IdentityType,
    scenario: IdentityScenario,
    version: str,
    model_bundle_version: str,
    preprocessor_version: str,
    calibration_dataset_version: str,
    signal_weights: dict[str, float],
    required_signals: tuple[str, ...],
    review_margin: float,
    minimum_confidence: float,
    samples: Iterable[CalibrationSample],
    minimum_precision: float = 0.0,
    minimum_recall: float = 0.0,
) -> ThresholdCalibrationProfile:
    distinct_signals = tuple(sorted(set(required_signals)))
    if not distinct_signals:
        raise ValueError("required_signals must not be empty")
    if identity_type in {"PRODUCT", "LOGO"} and len(distinct_signals) < 2:
        raise ValueError("PRODUCT/LOGO calibration must require multiple independent signals")
    if not 0 <= minimum_confidence <= 1:
        raise ValueError("minimum_confidence must be between 0 and 1")
    if not 0 <= review_margin <= 100:
        raise ValueError("review_margin must be between 0 and 100")
    for signal in distinct_signals:
        if signal not in signal_weights:
            raise ValueError(f"required signal has no weight: {signal}")
    if not signal_weights or any(weight <= 0 for weight in signal_weights.values()):
        raise ValueError("signal weights must be positive")
    metrics = select_calibrated_threshold(
        samples,
        identity_type,
        scenario,
        minimum_precision=minimum_precision,
        minimum_recall=minimum_recall,
    )
    return ThresholdCalibrationProfile(
        profile_id=profile_id,
        organization_id=organization_id,
        identity_type=identity_type,
        scenario=scenario,
        version=version,
        threshold=metrics.threshold,
        review_floor=max(0.0, metrics.threshold - review_margin),
        minimum_confidence=minimum_confidence,
        signal_weights=dict(signal_weights),
        required_signals=distinct_signals,
        model_bundle_version=model_bundle_version,
        preprocessor_version=preprocessor_version,
        calibration_dataset_version=calibration_dataset_version,
        metrics=metrics,
    )
